// stores/chart-store.js
var Fluxxor = require("fluxxor")
var actions = require("actions")

var _       = require('lodash')

var log     = require('debug')('src:node_modules:stores:chart-store')


var ChartStore = Fluxxor.createStore({

  initialize: function () {
    this.timeout = setTimeout(function(){}, 0);
    this.bindActions(
      actions.constants.LINECHART.HIGHLIGHTGROUP, this.highlightGroup,
      actions.constants.LINECHART.UNHIGHLIGHTGROUP, this.unHighlightGroup,
      actions.constants.LINECHART.HIGHLIGHTDATASET, this.highlightDataset,
      actions.constants.LINECHART.UNHIGHLIGHTDATASET, this.unHighlightDataset
    )
    this.highlightGroups = []
    this.highlightDatasets = []
    this.locked = false
    this._ = _
  },
  reset : function(){
    this.highlightGroups = []
    this.highlightDatasets = []    
  } ,   
  highlightGroup : function(payload){
      clearTimeout(this.timeout)
      this.timeout = setTimeout(_.bind(function(){
        this.reset()
        this.emit("change")
      },this),2500)
      this.highlightGroups = [payload.group]
      this.emit("change") 
  },
  unHighlightGroup : function(){ 
//    log('un highlight')   
      this.highlightGroups = []
//      this.emit("change");   
  },
  highlightDataset : function(payload){
      clearTimeout(this.timeout)
      this.timeout = setTimeout(_.bind(function(){
        this.reset()
        this.emit("change")
      },this),2500)
      this.highlightDatasets = [payload.datasetId]
      this.emit("change") 
  },
  unHighlightDataset : function(){ 
//    log('un highlight')   
      this.highlightDatasets = []
//      this.emit("change");   
  },
  getHighlightGroups : function(){
    return this.highlightGroups
  },
  getHighlightDatasets : function(){
    return this.highlightDatasets
  }

});

module.exports = ChartStore;
