// stores/chart-store.js
var Fluxxor = require("fluxxor")
var actions = require("actions")

var _       = require('lodash')

var log     = require('debug')('src:node_modules:stores:chart-store')


var ChartStore = Fluxxor.createStore({

  initialize: function () {
    this.timeout = setTimeout(function(){}, 0);
    this.bindActions(
      actions.constants.LINECHART.HIGHLIGHTGROUP, this.highlightGroup,
      actions.constants.LINECHART.UNHIGHLIGHTGROUP, this.unHighlightGroup
    )
    this.highlightGroups = []
    this.locked = false
    this._ = _
  },
      
  highlightGroup : function(payload){
      clearTimeout(this.timeout)
      this.timeout = setTimeout(_.bind(function(){
        log('un highlight after timeout')  
        this.highlightGroups = []
        this.emit("change")
      },this),3000)
      this.highlightGroups = [payload.group]
      this.emit("change") 
  },
  unHighlightGroup : function(){ 
    log('un highlight')   
      this.highlightGroups = []
//      this.emit("change");   
  },
  getHighlightGroups : function(){
    return this.highlightGroups
  }

});

module.exports = ChartStore;
