// stores/dataset-store.js
var Fluxxor = require("fluxxor");
var actions = require("actions");

var _       = require('lodash')

var log     = require('debug')('stores:dataset-store')


var DatasetStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', data)
    this.data = data[this.name].elements
    this.loading = false
    this.emit('change')
  },
  isLoading : function(){
    return this.loading
  },

  // loadDataFailure: function (error) {
  //   //
  // },

  // match: object: -> array
  query: function (match) {
    //TODO more query methods
    log('match', match, this.data)
    if (!match) return  this.data
    return _.where(this.data, match)
  },

  //match: func, obj: -> datum
  findOne: function (match) {
    log('findOne', match, this.data)
    return _.find(this.data, match)
  },

  //filterFunc: function: -> array
  filter: function (filterFunc) {
    return _.filter(this.data, filterFunc)
  },

  //NOTE: methods used if routeStore waitsFor for ethnicityStore to update

  // getState: function () {
  //   return {
  //     selected: this.selected
  //   }
  // },

  // handleSelectEthnicity: function (payload) {
  //   this.selected = _.where(this.data, { id: payload.ethnicity })
  // }


});

module.exports = DatasetStore;
