// stores/data-store.js
//flux
var Fluxxor = require("fluxxor");

//helpers
var _       = require('lodash');
var log 		= require('debug')(':stores:data')

//utils
var sanitise  = require('utils/sanitise')

var DataStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    // call our load data closure with success callback    
    this.loadData({
      key:options.key, 
      bucket:options.bucket, 
      sheet:options.sheet, 
      isProxy:options.isProxy
    }, this.loadDataSuccess)()
    this.loading = true
  },
  // data load success callback 
  loadDataSuccess: function (data) {
    log('success', data)
    this.data = sanitise(data[this.name].elements)
    this.loading = false
    this.emit('change')
  },
  isLoading : function(){
    return this.loading
  },
  // loadDataFailure: function (error) {
  //   //
  // },
  query: function (match) {
    if (!match) return  this.data
    log('match', match)
    return _.clone(_.where(this.data, match),true)
  },

  //match: func, obj: -> datum
  findOne: function (match) {
    return _.clone(_.find(this.data, match),true)
  },


});

module.exports = DataStore;
