// stores/data-store.js
var Fluxxor = require("fluxxor");
var _       = require('lodash');
var log 		= require('debug')(':stores:data')
var sanitise  = require('utils/sanitise')


var DataStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', data)
    this.data = sanitise(data[this.name].elements)
    this.loading = false
    this.emit('change')
  },
  isLoading : function(){
    return this.loading
  },
  // loadDataFailure: function (error) {
  //   //
  // },
  query: function (match) {
    if (!match) return  this.data
    log('match', match)
    return _.clone(_.where(this.data, match),true)
  },

  //match: func, obj: -> datum
  findOne: function (match) {
    return _.clone(_.find(this.data, match),true)
  },


});

module.exports = DataStore;
