// stores/data-breakdown-store.js
var Fluxxor   = require("fluxxor")
var actions   = require("actions")
var Tabletop  = require('tabletop').Tabletop
var log       = require('debug')('stores:data-breakdown-store')
var _         = require('lodash')


var DataBreakdownStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', data)
    this.data = data[this.name].elements
    this.loading = false
    this.emit('change')
  },
  isLoading : function(){
    return this.loading
  },
  // loadDataFailure: function (error) {
  //   //
  // },

  query: function (match) {
    //TODO more query methods
    if (!match) return  this.data
    return _.where(this.data, match)
  },

  //match: func, obj: -> datum
  findOne: function (match) {
    return _.find(this.data, match)
  },

  //filterFunc: function(d): -> array
  filter: function (filterFunc) {
    return _.filter(this.data, filterFunc)
  },

  //NOTE: methods used if routeStore waitsFor for DataBreakdownStore to update

  // getState: function () {
  //   return {
  //     selected: this.selected
  //   }
  // },

  // handleSelectAge: function (payload) {
  //   this.selected = where(this.data, { id: payload.Age })
  // }


});

module.exports = DataBreakdownStore;
