// stores/indicator-store.js
var Fluxxor = require("fluxxor");
var actions = require("actions");
var each    = require('lodash.foreach');
var filter  = require('lodash.filter');
var where   = require('lodash.where');
var log     = require('debug')(':stores:indicator-store')
var slug    = require('slug')
slug.defaults.modegit  = 'rfc3986'


var IndicatorStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', data)
    this.data = data[this.name].elements
    this.makeSlugs()
    this.loading = false
    this.emit('change')
  },

  makeSlugs: function () {
    var space = / /g
    var comma = /,/g
    each(this.data, function (d) {
      d.slug = slug(d.title.replace(comma, '').split(space).slice(0, 4).join(' ')) //
      log('slug', d.slug)
    })
  },


  // loadDataFailure: function (error) {
  //   //
  // },

  query: function (match) {
    //TODO more query methods
    if (!match) return  this.data
    return where(this.data, match)
  }


});

module.exports = IndicatorStore;
