// stores/issue-store.js
var Fluxxor = require("fluxxor");
var actions = require("actions");
var each    = require('lodash.foreach')
var findOne = require('lodash.find')

var filter  = require('lodash.filter');
var where   = require('lodash.where');
var log     = require('debug')(':stores:issue-store')
var slug    = require('slug')
slug.defaults.mode = 'rfc3986'



var IssueStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.data = []
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', data)
    this.data = data[this.name].elements
    this.makeSlugs()
    this.loading = false
    this.emit('change')
  },
  isLoading : function(){
    return this.loading
  },
  makeSlugs: function () {
    var space = / /g
    var comma = /,/g
    each(this.data, function (d) {
      d.slug = slug(d.title.replace(comma, '').split(space).slice(0, 4).join(' ')) //
      log('slug', d.slug)

    })
  },


  // loadDataFailure: function (error) {
  //   //
  // },

  //match: func, obj: -> datum
  findOne: function (match) {
    return findOne(this.data, match)
  },


  query: function (match) {
    //TODO more query methods
    log('match', match)
    if (!match) return  this.data
    return where(this.data, match)
  }

  //NOTE: methods used if routeStore waitsFor for ethnicityStore to update

  // getState: function () {
  //   return {
  //     selected: this.selected
  //   }
  // },

  // handleSelectEthnicity: function (payload) {
  //   this.selected = where(this.data, { id: payload.ethnicity })
  // }


});

module.exports = IssueStore;
