// components/nav-carousel/nav-carousel.js
var React         = require('react')

//helpers
var _         = require('lodash')

var log       = require('debug')('src:components:nav-carousel')



module.exports = React.createClass({

  propTypes: {
    nextSlug: React.PropTypes.string,
    previousSlug: React.PropTypes.string,
    indicatorSlug: React.PropTypes.string,
    flux: React.PropTypes.object
  },

  render: function () {
    log('props', this.props)

    //display table
      //display table row
        //display table cell

    return (
      <div className="hrc-nav-carousel pull-right">
        <div className="hrc-nav-carousel-labels hrc-nav-carousel-row">
          <div className="small hrc-nav-carousel-label" >
            <a href='#' onClick={this.handleIndicatorClick(this.props.previousSlug)} >Previous</a>
          </div>
          <div className="small hrc-nav-carousel-label" >
            <a href='#' onClick={this.handleIndicatorClick(this.props.nextSlug)}>Next</a>
          </div>
          <div className="small hrc-nav-carousel-label" >
            <a href='#' onClick={this.handleIssueClick(this.props.issueSlug)}>Back</a>
          </div>
        </div>
        <div className="hrc-nav-carousel-icons hrc-nav-carousel-row">
          <div className="hrc-nav-carousel-cell" >
            <a href='#' onClick={this.handleIndicatorClick(this.props.previousSlug)} className="icon-previous"><div className='divider-right'></div></a>
          </div>
          <div className="hrc-nav-carousel-cell" >
            <a href='#' onClick={this.handleIndicatorClick(this.props.nextSlug)} className="icon-next"></a>
          </div>
          <div className="hrc-nav-carousel-cell">
            <a href='#' onClick={this.handleIssueClick(this.props.issueSlug)} className="icon-close"></a>
          </div>
        </div>
      </div>


    )
  },

  handleIssueClick: function (issueSlug) {
    log('init', issueSlug)
    var self = this
    return function (e) {
      log(e)
      self.props.flux.actions.issue.select(issueSlug)
    }
  },

  handleIndicatorClick: function (indicatorSlug) {
    log('init', indicatorSlug)
    var self = this
    return function (e) {
      log(e)
      self.props.flux.actions.indicator.select(indicatorSlug)
    }
  }


})
