// components/indicator/filters/filter-dataset.js
//main
var React       = require('react')

//components
var Input      	= require('react-bootstrap/lib/Input')
var Col         = require('react-bootstrap/lib/Col')
var Row         = require('react-bootstrap/lib/Row')
var Button         = require('react-bootstrap/lib/Button')


//helpers
var _ 					= require('lodash')

//logging
var log       	= require('debug')('components:indicator:time-select')

var style = {
	filterControl: {
		marginLeft: 20
	},
}

module.exports = React.createClass({

	propTypes : {
		flux: React.PropTypes.object,
		indicator: React.PropTypes.object,
		dateOptions: React.PropTypes.array
	},

  getInitialState: function () {
    var optionsMap = {}
    _.each(this.props.dateOptions, function (option) {
      optionsMap[option.value] = option
    })

    return {
      optionsMap: optionsMap
    }
  },

  componentWillReceiveProps: function (nextProps) {
    var optionsMap = {}
    _.each(nextProps.dateOptions, function (option) {
      optionsMap[option.value] = option
    })

    this.setState({ optionsMap: optionsMap })
  },

  render: function () {
		log('props', this.props, this.state)
		var flux = this.props.flux
		var dateClassName
    var DateComponent
    var ViewSelect
    var TimeLineSelect
    
    var dateOptions = this.props.dateOptions
    var selectedOption = _.filter(dateOptions,{active:true})
		

		if (dateOptions.length > 1) {
			dateClassName = 'date-dropdown'
			DateComponent = (
				<Input
          style={{ width: 85 }}
          type='select'
					name='indicator-date-dropdown'
					value={selectedOption.value}
					onChange={this.handleDatasetSelect} >
          {
            _.map(dateOptions, function (option) {
              return (
                <option value={option.value}>{option.label}</option>
              )
            })
          }
        </Input>
			)
		}
		else {
			dateClassName = 'date-header'
			DateComponent = <h3>{ this.props.dateOptions[0] ? this.props.dateOptions[0].label : null }</h3>
		}

    if (this.props.dateOptions.length > 1 && this.props.indicator.datatype !== 'categorical') {
      ViewSelect = (
        <Col md={2} style={{ float: 'right' }}>
          <Button onClick={this.handleViewSelect('snapshot')}>Snapshot</Button>
          <Button onClick={this.handleViewSelect('timeline')}>Over time</Button>
        </Col>
      )      
             
    }
    else {
      ViewSelect = (<Col md={2}/>)
    }



    return (
      <Row>
        <Col md={1}>
          {DateComponent}
        </Col>
        {ViewSelect}
      </Row>
    )

    DateComponent
  },

  handleDatasetSelect: function (e) {
    log('e', e.target.value)
    var datasetId = this.state.optionsMap[e.target.value].dataset.id
    this.props.flux.actions.indicator.selectFilter('dataset', datasetId)
  },

  handleViewSelect: function (view) {
    var self = this
    return function (e) {
      e.preventDefault()
      if (view === 'timeline') {
        self.props.flux.actions.indicator.selectFilter('dataset', 'all')
      } else {
        self.props.flux.actions.indicator.selectFilter('dataset', self.props.dateOptions[0].value)
      }
    }
  }



})
