// components/select-group/index.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var FluxMixin   = Fluxxor.FluxMixin(React)
var Select      = require('react-select')

var log       = require('debug')('src:components:select-group:index')

module.exports = React.createClass({

  propTypes: {
    flux: React.PropTypes.func.isRequired,
    groupType: React.PropTypes.string.isRequired
  },

  mixins: [
    FluxMixin
  ],

  render: function () {
    var groupTypes    = this.props.groupType
    var flux          = this.getFlux()

    var params        = flux.stores.route.getCurrentParams() //not used
    var query         = flux.stores.route.getCurrentQuery()
    //on render fetch latest data
    var data          = flux.stores[groupType].query(params)

    var defaultValue  = flux.stores[groupType].query({ id: 'all' }).label
    var value         = query[groupType] || defaultValue

    var options       = data.map(function (d) {
      return {
        value: d.id,
        label: d.label
      }
    })    

    log('data', data)
    log('params', params)
    log('props', this.props)

    return (
      <Select
        name={'select-'+groupType}
        // multi-select disabled for time being
        // multi={true}
        value={value}
        options={options}
        onChange={this.handleOnChange} />

    )
  },

  handleOnChange: function (val) {
    var currentQuery    = this.getFlux().stores.route.getCurrentQuery()
    var selectedValues  = (val.indexOf(',') !== -1) ? val.split(',') : [val]
    var groupType       = this.props.groupType

    //TODO more generic/decoupled in action call? 
    this.getFlux().actions.group.selectChange(groupType, selectedValues, currentQuery)
  }
})
