// components/issue/index.js
var React         = require('react')
var Fluxxor       = require('fluxxor')
var FluxMixin     = Fluxxor.FluxMixin(React)
var Router        = require('react-router')

var RouteHandler  = Router.RouteHandler

//helpers
var pluck         = require('lodash.pluck')
var where         = require('lodash.where')
var find          = require('lodash.find')
var result        = require('lodash.result')
var capitalize    = require('capitalize')

var log       = require('debug')('src:components:issue')


module.exports = React.createClass({

  propTypes: {
    flux: React.PropTypes.object
  },

  getInitialState: function () {
    var flux = this.props.flux
    var router = flux.stores.routes
    var issueName = capitalize(router.getCurrentRoute().substring(1))

    var issueData = flux.stores.issues.query({ title: issueName })[0]
    var indicators = flux.stores.indicators.query({ issue: issueData.id })

    return {
      issueData: issueData,
      indicators: indicators,
      issueName: issueName
    }
  },

  render: function () {

    return (
      <div className='issue'>
        <h3>Issue</h3>
        <h2>{this.state.issueName}</h2>
        <p>{this.state.issueData.commentary}</p>
          {this.state.indicators.map(function (indicator) {
            return (
              <button onClick={this.handleIndicatorClick(indicator.id)}>{indicator.title}</button>
            )
          }.bind(this))}
      </div>
    )
  },

  componentWillReceiveProps: function (nextProps) {
    var flux = nextProps.flux
    var router = flux.stores.routes
    var issueName = capitalize(router.getCurrentRoute().substring(1))

    var issueData = flux.stores.issues.query({ title: issueName })[0]
    debug('issueData', issueData)
    var indicators = flux.stores.indicators.query({ issue: issueData.id })

    this.setState({
      issueData: issueData,
      indicators: indicators,
      issueName: issueName
    })

  },

  handleIndicatorClick: function (indicatorId) {
    var router = this.props.flux.stores.routes
    return function (e) {

      var pathName = '/' + [this.state.issueName.toLowerCase(), 'indicators', indicatorId].join('/')

      log('navigate to indicator', pathName)
      router.handlePathTransition(pathName)
    }.bind(this)
  }
})
