// components/issue/indicator-preview-range.js
var React           = require('react')

//helpers

var log       = require('debug')('src:components:issue:indicator-preview')

module.exports = React.createClass({


  render: function () {    
    var data = this.props.data
    
    var wrapStyle = { 
      position: 'relative', 
      paddingLeft : 0, 
      paddingRight: 0, 
      paddingTop : 5, 
      paddingBottom : 5,
      marginBottom : 8,
      paddingBottom : 5,
      height:'30px',
      borderLeft: '2px solid black'}
    
    var barStyleTop = { 
      width: data.top.length + '%', 
      backgroundColor: data.top.colour}
    var barStyleBottom = { 
      width: data.bottom.length + '%', 
      backgroundColor: data.bottom.colour}
    var threshold = 75    
    var labelStyleTop = { 
      position: 'relative', 
      left: '100%', 
      marginLeft: 4, 
      color:data.top.colour }		
    if (data.top.length > threshold) {
      labelStyleTop = { 
        display:"block",
        position: 'relative', 
        left: 0, 
        marginLeft: 0, 
        color:'#fff', 
        textAlign:'right',
        backgroundColor:'transparent',
        width:'100%' }		
    }
    
    var labelStyleBottom = { 
      position: 'relative', 
      left: '100%', 
      marginLeft: 4, 
      color:data.bottom.colour }		    
    if (data.bottom.length > threshold) {
      labelStyleBottom = { 
        display:"block",
        position: 'relative', 
        left: 0, 
        marginLeft: 0, 
        color:'#fff', 
        textAlign:'right',
        backgroundColor:'transparent',
        width:'100%' }		
    }
		
    
    return (
      <div>      
        <div className={"top-group-label"} style={{color:data.top.colour}}>
          <div>
            <strong>Least vulnerable group, {this.props.date}</strong>
          </div>
          <div>
            {data.top.group}
          </div>
        </div>
        <div className="preview-chart" style={{position:'relative'}}>
          <div className='horizontal-bar-wrap'style={wrapStyle}>
            <div className='horizontal-bar'style={barStyleTop}>
              <span className='horizontal-bar-label' style={labelStyleTop}>{data.top.valueLabel}</span>
            </div>
          </div>        
          {
            (typeof data.ref !== 'undefined')
            ? <div className={"reference-group"}>
                <div className='vertical-line-wrap'style={{display:'block',position:'absolute',height:'68px',top:0,width: '100%',zIndex: 9}}>
                  <div className='vertical-line'style={{position:'relative',left:data.ref.length+'%',height:'68px',borderLeft:'1px solid black'}}>
                    <span className='vertical-line-label' style={{
                      color:'black',                      
                      height: '68px',
                      display: 'table-cell',
                      verticalAlign: 'middle',
                      fontSize: '11px',
                      paddingLeft: '4px'             
                    }} >
                      <span style={{fontWeight:'bold',paddingRight:'5'}}>{data.ref.valueLabel}</span>
                      <span>{data.ref.group}</span>
                    </span>
                  </div>
                </div>
              </div>
            : <div/> 
          }                
          <div className='horizontal-bar-wrap'style={wrapStyle}>
            <div className='horizontal-bar'style={barStyleBottom}>
              <span className='horizontal-bar-label' style={labelStyleBottom}>{data.bottom.valueLabel}</span>
            </div>
          </div>
        </div>
        <div className={"bottom-group"} style={{color:data.bottom.colour}}>
          <div>
            <strong>Most vulnerable group, {this.props.date}</strong>
          </div>
          <div>
            {data.bottom.group}
          </div>
        </div>
      
      </div>
    ) 
  },
 
 
})
