// components/issue/indicator-preview-range.js
var React           = require('react')

//helpers

// var log       = require('debug')('src:components:issue:indicator-preview')

module.exports = React.createClass({

  propTypes: {
    data: React.PropTypes.object,
    date: React.PropTypes.string
  },

  render: function () {
    var data = this.props.data
    var threshold = 75 
    
    var barStyleTop = { 
      width: data.top.length + '%', 
      backgroundColor: data.top.color}
    var barStyleBottom = { 
      width: data.bottom.length + '%', 
      backgroundColor: data.bottom.color}
    
    var labelStyleTop = { color:data.top.color }		
    var labelClassTop = 'horizontal-bar-label align-outside' 
    if (data.top.length > threshold) {
      labelClassTop = 'horizontal-bar-label align-inside' 
      labelStyleTop = { 
        backgroundColor:'transparent'    
      }
    } 
    var labelClassBottom = 'horizontal-bar-label align-outside'
    var labelStyleBottom = { color:data.bottom.color }    
    if (data.bottom.length > threshold) {
      labelClassBottom = 'horizontal-bar-label align-inside'      
      labelStyleBottom = { 
        backgroundColor:'transparent'    
      }
    } 


    return (
      <div className='preview-tile-range'>      
        <div className="group-label top-group-label" >
          <div className="group-label-title" style={{color:data.top.color}}>
            Least marginalised group, {this.props.date}
          </div>
          <div className='group-label-group'>
            {data.top.group}
          </div>
        </div>

        <div className="preview-chart">
          <div className='horizontal-bar-wrap horizontal-bar-wrap-top'>
            <div className='horizontal-bar-bg'>                                    
              <div className='horizontal-bar'style={barStyleTop}>
                <span className={labelClassTop} style={labelStyleTop}>
                  {data.top.valueLabel}
                </span>
              </div>
            </div>
          </div>
  
          {
            (typeof data.ref !== 'undefined')
            ? <div className={"reference-group"}>
                <div className='vertical-line-wrap'>
                  <div className='vertical-line'style={{left:data.ref.length+'%'}}>
                    <span className='vertical-line-label'>
                      <span className='vertical-line-label-group'>{data.ref.group}</span>
                    </span>
                  </div>
                </div>
              </div>
            : <div/>
          }
  
          <div className='horizontal-bar-wrap horizontal-bar-wrap-bottom'>
            <div className='horizontal-bar-bg'>                          
              <div className='horizontal-bar' style={barStyleBottom}>              
                <span className={labelClassBottom} style={labelStyleBottom}>
                  {data.bottom.valueLabel}
                </span>
              </div>
            </div>
          </div>
  
        </div>

        <div className="group-label bottom-group-label" >
          <div className="group-label-title" style={{color:data.bottom.color}}>
            Most marginalised group, {this.props.date}
          </div>
          <div className='group-label-group'>
            {data.bottom.group}
          </div>
        </div>
      </div>
    )
  },


})
