// components/issue/issue.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var NavPrimary      = require('components/nav-primary') 

//helpers

var _               = require('lodash')
var capitalize      = require('capitalize')
var extend          = require('xtend')

var log       = require('debug')('src:components:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'indicators')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      issuesLoading: flux.store('issues').isLoading(),
      indicatorsLoading : flux.store("indicators").isLoading(),
    }
  },

  render: function () {
    log('props, state', this.props, this.state)
    var flux = this.getFlux()
    var IssueView
    var IndicatorsView
    var issue
    var indicators

    if (this.state.issuesLoading) {
      IssueView = <h2>Loading...</h2>
    }
    else {
      issue = flux.stores.issues.findOne({ slug: this.props.params.issueSlug })

      IssueView = (
        <div>
          <h2>{issue.title}</h2>
          <p>{issue.commentary}</p>
        </div>
      )
    }

    if (this.state.issuesLoading || this.state.indicatorsLoading) {
      IndicatorsView = <p>Loading...</p>
    }
    else {
      indicators = flux.stores.indicators.query({ issue: issue.id })

      IndicatorsView = (
        <div>
          {
            indicators.map(function (indicator) {
              log('indicator', indicator)
              return (
                <button onClick={this.handleIndicatorClick(indicator.slug)}>{indicator.title}</button>
              )
            }.bind(this))
          }
        </div>
      )
    }

    return (
      <div>
        <NavPrimary 
          label = { {top:'Select',bottom:'Issue'} }
          type = 'issue'
          items = { flux.stores.issues.data.map(function(item){
              return extend(item,{
                active : flux.stores.routes.getCurrentParams().issueSlug === item.slug,                
                subtitle : (this.state.indicatorsLoading) ? '' : flux.stores.indicators.query({ issue: item.id }).length + ' Indicators'
              })               
            }.bind(this))
          } />
        <div className='issue'>
          <div className='container'>
            <div>{IssueView}</div>
            <div>{IndicatorsView}</div>
           </div>
        </div>
      </div>
    )
  },

  componentWillReceiveProps: function (nextProps) {
    log('nextProps', nextProps)
  },

  handleIndicatorClick: function (indicatorSlug) {
    return function (e) {
      this.props.flux.actions.indicator.select(indicatorSlug)
    }.bind(this)
  }
})
