// components/footer/footer.js
//main
var React       = require('react')

//helpers
var _ 					= require('lodash')

var Grid            = require('react-bootstrap/lib/Grid')
var Nav             = require('react-bootstrap/lib/Nav')
var Navbar          = require('react-bootstrap/lib/Navbar')
var NavItem 				= require('react-bootstrap/lib/NavItem')


//logging
var log       	= require('debug')('components:header')

var Brand = (
  <div className='header-logo'>
    <a className="icon-logo icon-big" href="http://hrc.co.nz" target="_blank"></a>
  </div>
)


module.exports = React.createClass({

  propTypes: {
    infoText: React.PropTypes.string,
    infoEmail: React.PropTypes.string,
    dumparkLink: React.PropTypes.string,
    parentSiteText: React.PropTypes.string,
    perentSiteLink: React.PropTypes.string,
    parentSiteURL: React.PropTypes.string,
  },

  render: function () {
		log('props', this.props)

		return (
			<footer className="hrc-footer">
        <ul className="list-inline container">
          <li className='hrc-footer-info hrc-footer-item'>
            <p>{this.props.infoText}<a href={'mailto:'+this.props.infoEmail}>{this.props.infoEmail}</a></p>
            <a className="dumpark-link small" href="http://dumpark.com" target="_blank">{this.props.dumparkLink}</a>
          </li>
          <li className='hrc-footer-parentsite hrc-footer-item'>
            <p>{this.props.parentSiteText}</p>
            <h4><a href={this.props.parentSiteURL} target="_blank">{this.props.parentSiteAnchorText}</a></h4>
          </li>
          <li className="hrc-footer-brand hrc-footer-item">
            <div className="hrc-footer-brand-logo">
              <a href={this.props.parentSiteURL} className="icon-hrc-logo-claim" target="_blank"></a>
            </div>
          </li>
        </ul>
      </footer>

    )
  },

  handleNavClick: function (page) {
    var flux = this.props.flux

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  }


})
