// components/footer/footer.js
//main
var React       = require('react')

var Grid        = require('react-bootstrap/lib/Grid')
var Row         = require('react-bootstrap/lib/Row')
var Col         = require('react-bootstrap/lib/Col')

//logging
var log         = require('debug')('components:footer')

module.exports = React.createClass({

  propTypes: {
    shouldHide: React.PropTypes.bool,
    infoText: React.PropTypes.string,
    infoEmail: React.PropTypes.string,
    dumparkLink: React.PropTypes.string,
    parentSiteText: React.PropTypes.string,
    perentSiteLink: React.PropTypes.string,
    parentSiteURL: React.PropTypes.string,
  },

  render: function () {
//    log('props', this.props)

		return (
      <footer className={this.props.shouldHide ? "hide" : "hrc-footer"}>
  			<Grid>
          <Row className="hrc-footer-row">
            <Col md={4} xs={4} className='hrc-footer-info hrc-footer-item'>
              <p>{this.props.infoText}<a href={'mailto:'+this.props.infoEmail}>{this.props.infoEmail}</a></p>
              <a className="builtby small" href="http://dumpark.com" target="_blank">
                interactive by <strong>dumpark</strong>
              </a>
            </Col>
            <Col md={4}  xs={4} className='hrc-footer-parentsite hrc-footer-item'>
              <p>{this.props.parentSiteText}</p>
              <h4><a href={this.props.parentSiteURL} target="_blank">{this.props.parentSiteAnchorText}</a></h4>
            </Col>
            <Col md={4}  xs={4} className="hrc-footer-brand hrc-footer-item">
              <div className="hrc-footer-brand-logo">
                <a href={this.props.parentSiteURL} className="icon-hrc-logo-claim" target="_blank"></a>
              </div>
            </Col>
          </Row>
        </Grid>
      </footer>

    )
  },

  handleNavClick: function (page) {
    var flux = this.props.flux

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  }


})
