// components/indicator/charts/chartSnapshotHeader.js
//main
var React       = require('react')

var Select       = require('react-select')

//helpers

var _           = require('lodash')
var capitalize   = require('capitalize')

//logging
var log         = require('debug')('components:indicator:table-header')

module.exports = React.createClass({

  propTypes : {
    flux            : React.PropTypes.object,
    breakdowns      : React.PropTypes.array,
    sortOptions     : React.PropTypes.array,
    sortBy          : React.PropTypes.string,
    referenceDatum  : React.PropTypes.object,
    dataType        : React.PropTypes.string
  },


  render: function () {
    log('props', this.props)


    return (
      <thead className='snapshot'>
        <tr className='table-header-groups'>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header' >{bd.title}</td>)
          })}
          <td className={this.props.dataType === 'categorical' ? 'hide' : 'sort-by-control'}>
            {
              this.renderSortByControls()
            }
          </td>
        </tr>
        <tr>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header-spacer' ></td>)
          })}
          <td>TODO baseline value and All people</td>
        </tr>
      </thead>

  )
  },

  renderSortByControls: function () {
    return ['Order by: ' ].concat(
        _.map(this.props.sortOptions, function (option, i, sortOptions) {
          var text = (this.props.sortBy === option) ? [<strong>{capitalize(option)}</strong>] : [capitalize(option)]
          if (i < sortOptions.length-1) text.push(' | ')
          return <a onClick={this.handleSortClick(option)} href="#">{text}</a>
        }, this)
      )
  },

  handleSortClick: function (value) {
    var self = this
    return function (e) {
      e.preventDefault()
      self.props.flux.actions.indicator.sortBy(value)
    }
  },


})
