// components/indicator/charts/chartSnapshotCell.js
//main
var React       = require('react')

//helpers

var _ 					= require('lodash')

//logging
var log       	= require('debug')('components:indicator:table-row')

module.exports = React.createClass({

	propTypes : {
    color: React.PropTypes.string,
    className: React.PropTypes.string,
    onClick: React.PropTypes.func,
    label: React.PropTypes.string,
    type: React.PropTypes.object,
    length: React.PropTypes.number,
	},

  render: function () {
		// log('props', this.props)
    var View

    if (this.props.type === 'horizontal-bar') {
      View = this.renderHorizontalBar()
    }
    else {
      View = this.renderText()
    }

		return View
  },

  renderHorizontalBar: function () {
    var barStyle = { width: this.props.length + '%', backgroundColor: this.props.color, color: '#333' }
		var labelStyle = { position: 'relative', left: '100%', paddingLeft: 3 }

    return (
      <td style={{paddingRight: 66}}
        className={this.props.className}
        onClick={this.props.onClick}>
        <div className='cell horizontal-bar'style={barStyle}>
					<span className='cell horizontal-bar-label' style={labelStyle}>{this.props.label}</span>
				</div>
			</td>
    )
  },

  renderText: function () {
		return (
			<td
        className={this.props.className}
        onClick={this.props.onClick}>
        { this.props.label }
			</td>
    )
  }
})
