// components/indicator/charts/chartSnapshotHeader.js
//main
var React         = require('react')

var Select         = require('react-select')

//helpers

var _             = require('lodash')
var makeDataLabel = require('utils/makeDataLabel')

//logging
var log           = require('debug')('components:indicator:table-header')

module.exports = React.createClass({

  propTypes : {
    flux            : React.PropTypes.object,
    breakdowns      : React.PropTypes.array,
    sortOptions     : React.PropTypes.array,
    sortBy          : React.PropTypes.string,
    refDatumLabel    : React.PropTypes.string,
    refLength       : React.PropTypes.number,
    refLabel        : React.PropTypes.string,
    unit             : React.PropTypes.string
  },


  render: function () {
    log('props', this.props)


    return (
      <thead className='snapshot'>
        <tr className='table-header-groups'>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header' >{bd.title}</td>)
          })}
          <td className={this.props.dataType === 'categorical' ? 'hide' : 'sort-by-control'}>
            {
              this.renderSortByControls()
            }
          </td>
        </tr>
        <tr>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header-spacer' ></td>)
          })}
          <td style={{paddingLeft: 0, paddingRight: 66}}>
            {
              [
                <span className='small' style={{float: 'left', width: this.props.refLength+'%'}}>{0 + this.props.unit}</span>,
                <span className='small' style={{}}>{this.props.refDatumLabel + ' ' + this.props.refLabel }</span>
              ]
            }
          </td>
        </tr>
      </thead>
  )
  },

  renderSortByControls: function () {
    return ['Order by: ' ].concat(
        _.map(this.props.sortOptions, function (option, i, sortOptions) {
          var text = (this.props.sortBy === option) ? [<strong>{option}</strong>] : [option]
          if (i < sortOptions.length-1) text.push(' | ')
          return <a onClick={this.handleSortClick(option)} href="#">{text}</a>
        }, this)
      )
  },

  handleSortClick: function (value) {
    var self = this
    return function (e) {
      e.preventDefault()
      self.props.flux.actions.indicator.sortBy(value)
    }
  },


})
