// components/indicator/charts/chartSnapshotHeader.js
//main
var React       = require('react')

var Select       = require('react-select')

//helpers

var _           = require('lodash')

//logging
var log         = require('debug')('components:indicator:table-header')

module.exports = React.createClass({

  propTypes : {
    flux            : React.PropTypes.object,
    breakdowns      : React.PropTypes.array,
    sortOptions     : React.PropTypes.array,
    sortBy          : React.PropTypes.string,
    referenceDatum  : React.PropTypes.object,
    dataType        : React.PropTypes.string
  },


  render: function () {
    log('props', this.props)


    return (
      <thead className='snapshot'>
        <tr className='table-header-groups'>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header' >{bd.title}</td>)
          })}
          <td className={this.props.dataType === 'categorical' ? 'hide' : 'sort-by-control'}>Order by
            {
              [
                <a onClick={this.handleSortClick('groups')} href="#">Group</a>,
                " | ",
                <a onClick={this.handleSortClick('value')} href="#">Value</a>
              ]
            }
          </td>
        </tr>
        <tr>
          {_.map(this.props.breakdowns, function (bd) {
            return (<td className='breakdown-column-header-spacer' ></td>)
          })}
          <td>TODO baseline value and All people</td>
        </tr>
      </thead>

  )


  },

  handleSortClick: function (value) {
    var self = this
    return function (e) {
      e.preventDefault()
      self.props.flux.actions.indicator.sortBy(value)
    }
  },


})
