// components/share-modal/share-modal.js
//main
var React       = require('react')

var Grid            = require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//logging
var log       	= require('debug')('components:share-modal')

module.exports = React.createClass({

  propTypes: {
    currentHREF: React.PropTypes.string,
    shareTitle: React.PropTypes.string,
    shouldHide: React.PropTypes.bool,
    toggleShareModal: React.PropTypes.func
  },

  render: function () {
		log('props', this.props)
    var className = this.props.shouldHide ? 'share-modal hide' : 'share-modal'



		return (
      
      <div className={className}>
        <div className='close-modal' onClick={this.handleCloseClick()}/>
        <div className='share-modal-inner'>
        <Grid>
          <Row className="share-modal-row">
            <Col md={5} xs={5}>
              <h1>Share this tool</h1>
              <h5>On social media</h5>
              <div className="share-modal-control">
                <a className='btn btn-share-modal share-modal-social'  onClick={this.followLink()} href={this.props.facebookShare} target="_blank">
                  <span className="icon-social-facebook"></span>
                  Facebook
                </a>
                <a className='btn btn-share-modal share-modal-social' href={this.props.twitterShare} target="_blank">
                  <span className="icon-social-twitter"></span>
                  Twitter
                </a>
              </div>
              <div className="share-modal-control">      
                <a className='btn btn-share-modal share-modal-social' href={this.props.linkedinShare} target="_blank">
                  <span className="icon-social-linkedin"></span>
                  LinkedIn
                </a>
              </div>
            </Col>
            <Col md={5} xs={5}>
              <h1 className="share-modal-header">Share current view</h1>
              <h5>Copy link</h5>
              <div className="share-modal-control">
                <input onClick={this.handleUrlInputClick} className="share-modal-link" type='text' value={this.props.currentHREF} readOnly>{}</input>
              </div>
            </Col>
            <div className="share-modal-close">
              <a href="#" className='btn btn-bg-white '  onClick={this.handleCloseClick()}  title="Close share modal">
                <div className="small hrc-nav-carousel-label" >
                  Back
                </div>
                <div className="hrc-carousel-icon hrc-carousel-icon-close">
                  <div className="icon-close"></div>
                </div>
              </a>
            </div>
          </Row>
        </Grid>
      </div>
      </div>
    )
  },

  handleUrlInputClick: function (e) {
    var target = e.target
    target.focus();
    target.select();

    setTimeout(function () {
      target.setSelectionRange(0, 9999);
    }, 1);
  },

  handleCloseClick: function () {
    return this.props.toggleShareModal()
  },
  stopPropagation: function () {
    return function (e) {
      e.preventDefault()
      e.stopPropagation()
    }
  },
  followLink: function () {
    return function (e) {

    }
  },
})
