// components/share-modal/share-modal.js
//main
var React       = require('react')

var Grid            = require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//logging
var log       	= require('debug')('components:share-modal')

module.exports = React.createClass({

  propTypes: {
    currentHREF: React.PropTypes.string,
    shareTweet: React.PropTypes.string,
    shareTitle: React.PropTypes.string,
    shouldHide: React.PropTypes.bool,
    toggleShareModal: React.PropTypes.func
  },

  render: function () {
		log('props', this.props)
    var className = this.props.shouldHide ? 'share-modal hide' : 'share-modal'

    var rootURI = window.location.protocol+'//'+window.location.host+'/'
    var encodedRootURI = encodeURIComponent(rootURI)
    var facebookShare = "https://www.facebook.com/sharer/sharer.php?u=" + encodedRootURI
    var twitterShare = "https://twitter.com/intent/tweet?status=" + encodeURIComponent(this.props.shareTweet) + " " + encodedRootURI
    var linkedinShare = "https://www.linkedin.com/shareArticle?mini=true&url=" + encodedRootURI + "&title=" + encodeURIComponent(this.props.shareTitle)

		return (
      <div className={className}>
        <div className='share-modal-inner'>
        <Grid>
          <Row className="share-modal-row">
            <Col md={5}>
              <h1>Share this tool</h1>
              <h5>On social media</h5>
              <div className="share-modal-control">
                <a className='btn btn-share-modal share-modal-social' href={facebookShare} target="_blank">
                  <span className="icon-social-facebook"></span>
                  Facebook
                </a>
                <a className='btn btn-share-modal share-modal-social' href={twitterShare} target="_blank">
                  <span className="icon-social-twitter"></span>
                  Twitter
                </a>
              </div>
                <a className='btn btn-share-modal share-modal-social' href={linkedinShare} target="_blank">
                  <span className="icon-social-linkedin"></span>
                  LinkedIn
                </a>
            </Col>
            <Col md={5}>
              <h1 className="share-modal-header">Share current view</h1>
              <h5>Copy link</h5>
              <div className="share-modal-control">
                <input className="share-modal-link" type='text' value={this.props.currentHREF}>{}</input>
              </div>
            </Col>
            <div className="share-modal-close">
              <a href="#" className='btn btn-bg-white '  onClick={this.handleCloseClick()}  title="Close share modal">
                <div className="small hrc-nav-carousel-label" >
                  Back
                </div>
                <div className="hrc-carousel-icon hrc-carousel-icon-close">
                  <div className="icon-close"></div>
                </div>
              </a>
            </div>
          </Row>
        </Grid>
      </div>
      </div>
    )
  },

  handleCloseClick: function () {
    return this.props.toggleShareModal()
  }
})
