// components/indicator/chartSnapshotHeader.js
//main
var React       = require('react')

var Select 			= require('react-select')

//helpers

var _ 					= require('lodash')

//logging
var log       	= require('debug')('components:indicator:table-header')

module.exports = React.createClass({

	propTypes : {
		flux            : React.PropTypes.object,
    dimensions      : React.PropTypes.array,
    sortOptions     : React.PropTypes.string,
    sortBy          : React.PropTypes.string,
    referenceDatum  : React.PropTypes.object
	},


  render: function () {
		log('props', this.props)
		var flux = this.props.flux



		return (
			<thead className='snapshot'>
				<tr className='table-header-groups'>
          {_.map(this.props.dimensions, function (dim) {
						return (<td className='breakdown-column-header' >{dim.title}</td>)
					})}
          <td>Sort by TODO sort select</td>
				</tr>
				<tr>
          {_.map(this.props.dimensions, function (dim) {
						return (<td className='breakdown-column-header-spacer' ></td>)
					})}					
          <td>TODO baseline value and All people</td>
				</tr>
			</thead>

  )


  },

	handleHeaderColumnClick: function (id) {
		return function () {
			//TODO
		}
	}


})
