// components/indicator/snapshot.js
//main
var React       = require('react')

//components
var Select      = require('react-select')

//helpers
var _ 					= require('lodash')

//logging
var log       	= require('debug')('components:indicator:snapshot')

module.exports = React.createClass({

	propTypes : {
		flux: React.PropTypes.object,
		params: React.PropTypes.object,
    data: React.PropTypes.array,
    datasets: React.PropTypes.array
	},

  render: function () {
		log('props', this.props)
		var flux = this.props.flux
    var DateComponent
    var options = _.map(this.props.datasets, function (dataset) {
      return {
        value: String(dataset['observation_date']),
        label: 'Year: '+ dataset['observation_date']
      }
    })

    log('options', options)
    //set DateComponent
		if (options.length > 1) {
			DateComponent = (
				<Select
					name='indicator-date-dropdown'
					value={options[0].value}
					options={options} //TODO remove initial value from options ?
					onChange={this.handleDateSelect} />
			)
		}
		else {
			DateComponent = <h3>{options[0].label}</h3>
		}

		return (
			<div className='snapshot'>
        <div className={options.length > 1 ? 'date-dropdown' : 'date-header'}>
          {DateComponent}
        </div>
			</div>

  )

		//Year: (most recent Year)

		//breakdown 1: breakdown 2:... Order:


		// var columns = Object.keys(data[0]).map(function (key) { return capitalize(key) })
		//
    // return (
		// 	<div className='indicator'>
		// 		<h2>{indicator.title}</h2>
		// 		<Table className='indicator-table table' columns={columns} data={data}/>
		// 	</div>
    // )

  },

	handleDateSelect: function () {
		//TODO
	}


})
