// components/indicator/table-header.js
//main
var React       = require('react')

var Select 			= require('react-select')

//helpers

var _ 					= require('lodash')

//logging
var log       	= require('debug')('components:indicator:table-header')

module.exports = React.createClass({

	propTypes : {
		flux: React.PropTypes.object,
    breakdowns: React.PropTypes.array,
    params: React.PropTypes.object,
		tableHeaderGroups: React.PropTypes.array
	},


  render: function () {
		log('props', this.props)
		var flux = this.props.flux



		return (
			<thead className='snapshot'>
				<tr className='table-header-groups'>
					{_.map(this.props.tableHeaderGroups, function (headerGroup) {
						return (<td className='table-header-group'>{headerGroup}</td>)
					})}
				</tr>
				<tr>
					{_.map(this.props.breakdowns, function (breakdown) {
						return (<td className='breakdown-column-header' >{breakdown.title}</td>)
					})}
				</tr>
			</thead>

  )


  },

	handleHeaderColumnClick: function (id) {
		return function () {
			//TODO
		}
	}


})
