// components/main/index.js
//main
var React           = require('react')
var Fluxxor         = require('fluxxor')

//react helpers
var FluxMixin       = Fluxxor.FluxMixin(React)

//components
var Select          = require('react-select')
var FilterGroup     = require('components/filter-group')
var IndicatorView   = require('components/indicator-view')

//debugging
var log             = require('debug')('src:components:main:index')

module.exports = React.createClass({

  mixins: [
    FluxMixin
  ],

  render: function () {
    var flux             = this.getFlux()
    var groupType        = 'ethnicity' //hardcoded for now
    var indicatorId      = 'unempl'
    var indicatorLabel   = 'Unemployment'

    var currentQuery     = flux.stores.route.getCurrentQuery()
    var currentEthnicity = currentQuery ? currentQuery[groupType] : null
    var groupLabel       = flux.stores[groupType].query({ id: currentEthnicity }).label || '?'
    var queryObject = {
      indicator: indicatorId,
      //other groupTypes hardcoded
      age: '',
      disability: '',
      sex: '',
      sogii: ''
    }
    queryObject[groupType] = currentEthnicity

    //using lodash.where -> array
    var datum             = flux.stores.data.query(queryObject)[0] || '?'

    log('groupLabel', groupLabel)
    log('value', datum)


    return (
      <div>
        <FilterGroup groupType={groupType} flux={flux} />
        <IndicatorView 
          id={'indicator-view-'+indicatorId} 
          header={indicatorLabel+'-'+groupLabel}
          datum={datum}  />
      </div>
    )

  },

})
