// components/multi-select/index.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var FluxMixin   = Fluxxor.FluxMixin(React)
var Select      = require('react-select')
var pluck       = require('lodash.pluck')
var _       = require('lodash')


var log       = require('debug')('src:components:multi-select:index')

module.exports = React.createClass({

  propTypes: {
    flux: React.PropTypes.func.isRequired,
    groupType: React.PropTypes.string.isRequired
  },

  mixins: [
    FluxMixin
  ],

  render: function () {
    var flux          = this.getFlux()
    var params        = flux.stores.route.getCurrentParams() //not used
    var query         = flux.stores.route.getCurrentQuery()

    var options       = _.map(this.props.options, function (option) {
      log('option', option)
      return {
        value: option,
        label: _.result(_.find(groupData, { id: option }), 'label')
      }
    })

    log('options', options)
    log('props', this.props)

    return (
      <Select
        name={'select-'+groupType}
        multi={true}
        value={value}
        options={options}
        onChange={this.handleOnChange} />

    )
  },

  handleOnChange: function (val) {
    var currentQuery    = this.getFlux().stores.route.getCurrentQuery()
    var selectedValues  = (val.indexOf(',') !== -1) ? val.split(',') : [val]

    log('selectedValues', selectedValues)

    //TODO more generic/decoupled in action call?
    this.getFlux().actions.group.selectChange(groupType, selectedValues, currentQuery)
  }
})
