// components/about/about.js
var React         = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')
var NavPrimary  = require('components/nav-primary')

var _           = require('lodash')

var log       = require('debug')('src:components:about')

module.exports = React.createClass({
  mixins: [
    StoreWatchMixin('indicators')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      indicatorsLoading : flux.store("indicators").isLoading()
    }
  },

  render: function () {
    log('props', this.props)
    var flux = this.props.flux

    return (
    <div>
      <NavPrimary
        flux = {flux}
        type = 'issue'
        items = {
          _.map(flux.stores.issues.data,function(item){
            return _.assign({},item,{
              active : flux.stores.routes.getCurrentParams().issueSlug === item.slug,
              subtitle : (this.state.indicatorsLoading) ? '' : flux.stores.indicators.query({ issue: item.id }).length + ' Indicators'
            })
          },this)
        }
        />
      <Grid className='about'>
        <Row>
          <Col md={9}>
            <div className='about-title'>
              <h2>About</h2>
            </div>
            <div className='about-lead'>
              <h1 className='about-lead-header'>The principle of equality underpins all human rights.
                It requires the establishment of a society in which all human beings will be accorded equal dignity and respect regardless of their membership of particular groups.
              </h1>
              <p className='about-lead-text'>
                The right to work is a fundamental human right, strongly established in international law. To paraphrase Eleanor Roosevelt human rights begin close to home, and must live in schools, homes, communities and workplaces. Work is not solely a source of income that provides for the basic necessities in life, but has the potential to satisfy social, intellectual and personal needs and therefore is integral for a life of human dignity.
              </p>
            </div>
            <div className="about-summary">
              <h1 className="about-summary-header">Tracking Equality at Work </h1>
              <p className='about-lead-text'>
                This web-based tool brings together a suite of employment data so that fairness and equality of outcomes at work between population groups can be compared. Data sets have been collected from a number of sources, some publically available, some data purchased and other data supplied on request. The commentary has been developed by the Human Rights Commission.   Further indicators and data sets will be added over time. The Human Rights Commission presents disaggregated data on four key aspects of work: Employment; Pay; Leadership and Discrimination. The interactive tool allows analysis of equality on any one of these issues by sex, ethnicity, age, disability, over time.
              </p>
              <p>
                This capability gives a clear picture of how groups are faring at work. The infographics illustrate gaps and the position of population groups in relation to each other. Where appropriate a disproportionality index (DPI)  is reported. This measure was developed to enable comparisons between different sized groups, in effect to quantify under and over-representation on an indicator.
              </p>
              <p>
                The web-tool makes it possible to track the persistence of inequality over time, and whether or not progress towards equality is being made. It is also makes it possible to track the outcomes of a particular group across multiple indicators. The web-based tool can also be used to disaggregate groups across several demographic characteristics.
              </p>
              <p>
                The persistence of inequalities over time and the consistency of inequality over many of the indicators for particular groups indicate the possibility of structural and systemic discrimination.  It is hoped that this data will be used to monitor progress on equality indicators at work and to provide an evidence base for action to achieve equality.
              </p>
            </div>
          </Col>
        </Row>
      </Grid>
      </div>
    )
  }
})
