// components/about/about.js
var React         = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')
var NavPrimary  = require('components/nav-primary')

var _           = require('lodash')

var log       = require('debug')('src:components:about')

module.exports = React.createClass({
  mixins: [
    StoreWatchMixin('indicators')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      indicatorsLoading : flux.store("indicators").isLoading()
    }
  },

  render: function () {
    log('props', this.props)
    var flux = this.props.flux

    return (
    <div className="page-wrapper page-wrapper-about">
      <NavPrimary
        flux = {flux}
        type = 'issue'
        items = {
          _.map(flux.stores.issues.data,function(item){
            return _.assign({},item,{
              active : flux.stores.routes.getCurrentParams().issueSlug === item.slug,
              subtitle : (this.state.indicatorsLoading) ? '' : flux.stores.indicators.query({ issue: item.id }).length + ' Indicators'
            })
          },this)
        }
        />
      <div className="page-content-about page-content">        
        <Grid className='page-section about-section'>
          <Row>
            <Col md={9} xs={9}>
              <div className='about-title'>
                <h2>About</h2>
              </div>
              <div className='about-lead'>
                <h3 className='about-lead-header'>
                  The Universal Declaration of Human Rights states that “Everyone has the right to work”. 
                  It also outlines that workers should be free from discrimination and have the right to equal pay for equal work. 
                </h3>
                <p className='about-lead-text'>
                  The right to work is a fundamental human right, strongly established in international law. 
                  To paraphrase Eleanor Roosevelt human rights begin close to home, and must live in schools, homes, communities and workplaces. 
                  Work is not solely a source of income that provides for the basic necessities in life, but has the potential to satisfy social, intellectual and personal needs and therefore is integral for a life of human dignity.
                </p>
              </div>
              <div className="about-summary">
                <h3 className="about-summary-header">Tracking Equality at Work </h3>
                <p className='about-lead-text'>
                  Tracking Equality at work is a web-based tool that brings together data relating to four key indicators Employment, Discrimination, Leadership and Pay in New Zealand from various sources.
                </p>
                <p>
                  The data is updated annually to provide a snapshot of how various groups based on gender, ethnicity, age, and disability are doing in each of the four key areas. 
                </p>
                <p>
                  The web-tool makes it possible to track the persistence of inequality over time, and whether or not progress towards equality is being made. 
                  It is also makes it possible to track the outcomes of a particular group across multiple indicators. 
                  The web-based tool can also be used to disaggregate groups across several demographic characteristics.
                </p>
              </div>
            </Col>
          </Row>
        </Grid>
      </div>
    </div>
    )
  }
})
