// components/filter-group/index.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var Router      = require('react-router')
var FluxMixin   = Fluxxor.FluxMixin(React)
var Select      = require('react-select')

var log       = require('debug')('src:components:filter-group:index')

module.exports = React.createClass({

  propTypes: {
    flux: React.PropTypes.func.isRequired,
    groupType: React.PropTypes.string.isRequired
  },

  contextTypes: {
    router: React.PropTypes.func.isRequired
  },

  mixins: [
    FluxMixin
  ],

  render: function () {
    var params  = this.context.router.getCurrentParams() //not used
    var query   = this.context.router.getCurrentQuery()
    var groupType = this.props.groupType
    //on render fetch latest data
    var data    = this.getFlux().stores[groupType].query(params)
    log('data', data)
    log('params', params)
    log('props', this.props)
    log('context', this.context)

    var value = query[groupType] || null

    var options = data.map(function (d) {
      return {
        value: d.id,
        label: d.label
      }
    })    

    return (
      <div>
        <Select
          name={'select-'+groupType}
          // multi-select disabled for time being
          // multi={true}
          value={value}
          options={options}
          onChange={this.handleOnChange} />
      </div>

    )

  },

  handleOnChange: function (val) {
    var currentQuery    = this.context.router.getCurrentQuery()
    var selectedValues  = (val.indexOf(',') !== -1) ? val.split(',') : [val]

    //TODO more generic/decoupled in action call? 
    this.getFlux().actions.group.selectChange(this.groupType, selectedValues, currentQuery)
  }
})
